### PR TITLE
Fix some debug info mangling bugs

### PR DESCRIPTION
**Pull request by slavapestov**
- rdar://problem/48004306
- rdar://problem/48014983
- rdar://problem/47980487

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 1]
```diff
@@ -742,6 +742,22 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     if (!Ty->hasTypeParameter())
       Ty = Ty->mapTypeOutOfContext();
 
+    // Strip off top level of type sugar (except for type aliases).
+    // We don't want Optional<T> and T? to get different debug types.
+    while (true) {
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp]: The control flow of this loop is a bit hard to follow.

**Review by eeckstein — CHANGES_REQUESTED** [review]
The code changes lgtm.
Can you please also update docs/ABI/Mangling.rst?

File: lib/IRGen/IRGenDebugInfo.cpp (hunk 1 above)
Comment (adrian-prantl) [reply lib/IRGen/IRGenDebugInfo.cpp]: But I also don't have a much better idea.

**Review by adrian-prantl — COMMENTED** [review]
Thanks!

**Comment by slavapestov** [comment]
@eeckstein Oops, I forgot! I'll do it in the next PR since this one is already running a lot of tests.

File: lib/IRGen/IRGenDebugInfo.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/IRGen/IRGenDebugInfo.cpp]: I think much of this code goes away altogether if we give up on supporting `-gdwarf-types`
